### PR TITLE
sprite fillbuffer cache & sprite misc change when set spf

### PR DESCRIPTION
**Pull request by LinYunMo**
Re: https://github.com/cocos-creator/3d-tasks/issues/7806

### CODE REVIEW

**Comment by pandamicro** [comment]
https://github.com/cocos-creator/3d-tasks/issues/7487

File: cocos/2d/assembler/sprite/simple.ts [hunk 1]
```diff
@@ -170,6 +142,42 @@ export const simple: IAssembler = {
...
+        // const commitBuffer: IUIRenderData = renderer.createUIRenderData();
+
+        let buffer = renderer.acquireBufferBatch()!;
+
+        let vertexOffset = buffer.byteOffset >> 2;
+        let indicesOffset = buffer.indicesOffset;
+        let vertexId = buffer.vertexOffset;
+
+        const isRecreate = buffer.request();
+        if (!isRecreate) {
```
Comment (pandamicro) [inline cocos/2d/assembler/sprite/simple.ts]: The naming  here  is confusing, should only update buffer if it is recreated, so I suggest name this variable `bufferUnchanged`

File: cocos/2d/components/sprite.ts [hunk 2]
```diff
@@ -763,8 +763,12 @@ export class Sprite extends Renderable2D {
 
         if (spriteFrame) {
             if (!oldFrame || spriteFrame !== oldFrame) {
-                if  (spriteFrame.loaded) {
-                    this._onTextureLoaded();
+                if (spriteFrame.loaded) {
+                    if (oldFrame && oldFrame.texture.getHash() !== spriteFrame.texture.getHash()) {
```
Comment (pandamicro) [inline cocos/2d/components/sprite.ts]: If old frame not exist, we should also invoke `_onTextureLoaded`, the logic here is too complicated

```
let textureChanged = false;
if (spriteFrame) {
    if (!oldFrame || oldFrame.texture !== spriteFrame.texture) {
         textureChanged = true;
    }
    if (spriteFrame.loaded) {
        if (textureChanged) {
            this._onTextureLoaded();
        }
        else {
            this._applySpriteSize();
        }
    }
    else {
        spriteFrame.once('load', this._onTextureLoaded, this);
    }
}
```

File: cocos/2d/framework/ui-transform.ts [hunk 3]
```diff
@@ -92,6 +92,7 @@ export class UITransform extends Component {
         } else {
             this.node.emit(SystemEventType.SIZE_CHANGED);
         }
+        this._rectDirty = true;
```
Comment (pandamicro) [inline cocos/2d/framework/ui-transform.ts]: should use the same mechanism to mark render data vertex dirty:

```
this.node.uiProps.uiComp.markForUpdateRenderData(true);
```

File: cocos/2d/assembler/sprite/simple.ts [hunk 4]
```diff
@@ -170,6 +142,42 @@ export const simple: IAssembler = {
...
         }
+    },
+
+    fillBuffers (sprite: Sprite, renderer: Batcher2D) {
+        if (sprite === null) {
+            return;
+        }
+
+        const vData = sprite.renderData!.vData!;
+        if (sprite.node.hasChangedFlags || (this.uiTransCache && this.uiTransCache._rectDirty)) {
```
Comment (pandamicro) [inline cocos/2d/assembler/sprite/simple.ts]: We can only check for `sprite.node.hasChangedFlags` here. UITransform size and anchor dirty should be updated directly in `updateVertexData`

File: cocos/2d/assembler/sprite/simple.ts [hunk 5]
```diff
@@ -170,6 +142,42 @@ export const simple: IAssembler = {
...
+            this.updateWorldVerts(sprite, vData);
+            this.uiTransCache._rectDirty = false;
+        }
```
Comment (pandamicro) [inline cocos/2d/assembler/sprite/simple.ts:156]:
```
if (sprite.node.hasChangedFlags) {
    this.updateWorldVerts(sprite, vData);
}
```

File: cocos/2d/assembler/sprite/simple.ts [hunk 6]
```diff
@@ -229,13 +237,12 @@ export const simple: IAssembler = {
...
         dataList[0].x = l;
         dataList[0].y = b;
-        dataList[0].z = 0;
 
         dataList[3].x = r;
         dataList[3].y = t;
-        dataList[3].z = 0;
 
         renderData.vertDirty = false;
+        this.uiTransCache = uiTrans;
```
Comment (pandamicro) [inline cocos/2d/assembler/sprite/simple.ts]: No need to save uiTransCache, just update world vertex here

```
this.updateWorldVerts(sprite, renderData.vData);
```

**Review by pandamicro — COMMENTED** [review]
I don't feel right about `rectDirty` design and the checks in fillBuffers, it feels like adding up complexity and easy to trigger issues

File: cocos/2d/assembler/sprite/simple.ts (hunk 5 above)
Comment (LinYunMo) [reply cocos/2d/assembler/sprite/simple.ts:156]: will be update twice，node.hasChangedFlags always true until frame finish

**Review by pandamicro — APPROVED** [review]

**Pull request merged**